docs/config.md
# Configuration file - config.json

```js
{
  "nodejs": {
    // (string) "development" or "production". Default is "production".
    "env": "production",

    // (string) Which IP to bind to? Default is "0.0.0.0"
    "host": "0.0.0.0",

    // (number) Which PORT to bind to? Default is 3000
    "port": 3000,

    // (boolean) Run server in https mode
    "ssl": false,

    // (string) Path to the SSL CA certificate
    // See SSL certificates section below for more details
    "sslCa": "ca.crt",

    // (string) Path to the SSL certificate
    // See SSL certificates section below for more details
    "sslCert": "server.crt",

    // (string) Path to the SSL privat key
    // See SSL certificates section below for more details
    "sslKey": "server.key"
  },

  "rcScanner": {
    "audio": {
      // Which audio device to use. -1 is for default device
      // You can get a list of your audio device with the
      // following command: 'npm server list-audio'
      "deviceId": -1,

      // Reconnect after this amount of milliseconds to the audio
      // device if the naudiodon library has crashed.
      "reconnectInterval": 2000,

      // The sampling rate that match your audio device
      "sampleRate": 44100,

      // PCM squelch value.
      // Below that value, server won't send audio data to client
      // Value of 0 will disable this feature.
      // You should try a value between 0 and 5000. It all depends
      // if you do have ground loop noises or not.
      "squelch": 100
    },

    "com": {
      // The baud rate of the port to be opened. This should match one of
      // the commonly available baud rates, such as 110, 300, 1200, 2400,
      // 4800, 9600, 14400, 19200, 38400, 57600, or 115200. Custom rates
      // are supported best effort per platform. The device connected to
      // the serial port is not guaranteed to support the requested baud
      // rate, even if the port itself supports that baud rate.
      "baudRate": 115200,

      // Must be one of these: 8, 7, 6, or 5.
      "dataBits": 8,

      // Must be one of these: 'none', 'even', 'mark', 'odd', 'space'.
      "parity": "none",

      // Use by some drivers; poll scanner status every x milliseconds.
      "pollingInterval": 250,

      // The system path of the serial port you want to open.
      // For example, /dev/ttyXXX0 on Mac/Linux, or COM1 on Windows.
      "port": "/dev/ttyUSB0",

      // Retry to connect to serial port after this amount of
      // milliseconds.
      "reconnectInterval": 2000,

      // flow control setting.
      "rtscts": false,

      // Must be one of these: 1 or 2.
      "stopBits": 1
    },

    // Some radio scanners display their serial number.
    // Here you can choose not to display the serial number
    // (replaced by zeros) on the client side.
    "hideSerialNumber": true,

    // One of the supported scanner model.
    // Use 'node server list-models' for the list of
    // supported models.
    "model": "bcd386t",

    "webSocket": {
      // Disconnect dead clients after x milliseconds.
      "keepAlive": 30000,

      // The client will try to reconnect to server's websockets
      // every x milliseconds.
      "reconnectInterval": 2000
    }
  }
}
```


CHANGELOG.md
# Version 0.10

- Add compatibility to model Uniden BCD386T.
- Configuration is now done through a config.json file. Make sure to delete your `.env` file after running the server at least once.

# Version 0.9

- First public beta version.


server/lib/rc-scanner/config.js
/*
 * *****************************************************************************
 * Copyright (C) 2019-2021 Chrystian Huot
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 * ****************************************************************************
 */

'use strict';

import { unknown } from './models.js';

export class Config {
    constructor(app) {
        const config = app.config.rcScanner;

        this.audio = {
            deviceId: typeof config?.audio?.deviceId === 'number'
                ? config.audio.deviceId
                : parseInt(process.env.RC_AUDIO_DEVICE_ID, 10) || -1,
            reconnectInterval: typeof config?.audio?.reconnectInterval === 'number'
                ? config.audio.reconnectInterval
                : parseInt(process.env.RC_AUDIO_RECONNECT_INTERVAL, 10) || 2000,
            sampleRate: typeof config?.audio?.sampleRate === 'number'
                ? config.audio.sampleRate
                : parseInt(process.env.RC_AUDIO_SAMPLE_RATE, 10) || 44100,
            squelch: typeof config?.audio?.squelch === 'number'
                ? config.audio.squelch
                : parseInt(process.env.RC_AUDIO_SQUELCH, 10) || 100,
        };

        this.com = {
            baudRate: typeof config?.com?.baudRate === 'number'
                ? config.com.baudRate
                : parseInt(process.env.RC_COM_BAUDRATE, 10) || 115200,
            dataBits: typeof config?.com?.dataBits === 'number'
                ? config.com.dataBits
                : parseInt(process.env.RC_COM_DATABITS, 10) || 8,
            parity: typeof config?.com?.parity === 'string'
                ? config.com.parity
                : process.env.RC_COM_PARITY || 'none',
            pollingInterval: typeof config?.com?.pollingInterval === 'number'
                ? config.com.pollingInterval
                : parseInt(process.env.RC_COM_POLLING_INTERVAL, 10) || 500,
            port: typeof config?.com?.port === 'string'
                ? config.com.port
                : process.env.RC_COM_PARITY || process.platform === 'win32' ? 'com1' : '/dev/ttyACM0',
            reconnectInterval: typeof config?.com?.reconnectInterval === 'number'
                ? config.com.reconnectInterval
                : parseInt(process.env.RC_RECONNECT_INTERVAL, 10) || 5000,
            rtscts: (config?.com?.rtscts || '').toLowerCase() === 'true'
                ? true
                : (process.env.RC_COM_RTSCTS || '').toLowerCase() === 'true' ? true : false,
            stopBits: typeof config?.com?.stopBits === 'number'
                ? config.com.stopBits
                : parseInt(process.env.RC_COM_STOPBITS, 10) || 1,
        };

        this.hideSerialNumber = typeof config?.hideSerialNumber === 'boolean'
            ? config.hideSerialNumber
            : (process.env.RC_HIDE_SERIAL_NUMBER || '').toLowerCase() === 'true' ? true : false;

        this.model = typeof config?.model === 'string' && config.model.length
            ? config.model.toLowerCase()
            : process.env.RC_MODEL || unknown;

        this.webSocket = {
            keepAlive: typeof config?.webSocket?.keepAlive === 'number'
                ? config.webSocket.keepAlive
                : parseInt(process.env.RC_WEBSOCKET_KEEP_ALIVE, 10) || 30000,
            reconnectInterval: typeof config?.webSocket?.reconnectInterval === 'number'
                ? config.webSocket.reconnectInterval
                : parseInt(process.env.RC_WEBSOCKET_RECONNECT_INTERVAL, 10) || 2000,
        };

        app.router.get('/config', (req, res) => {
            res.send({
                model: this.model,
                reconnectInterval: this.webSocket.reconnectInterval,
                sampleRate: this.audio.sampleRate,
            });
        });
    }
}
